templates/boltzmann/middleware/honeycomb.ts

```diff
@@ -9,77 +19,179 @@ import beeline from 'honeycomb-beeline'
   }
 
-  const schema = require('honeycomb-beeline/lib/schema')
-  const tracker = require('honeycomb-beeline/lib/async_tracker')
-
-  return function honeycombTrace (next: Handler) {
-    return (context: Context) => {
-      const traceContext = _getTraceContext(context)
-      const trace = beeline.startTrace({
-        [schema.EVENT_TYPE]: 'boltzmann',
-        [schema.PACKAGE_VERSION]: '1.0.0',
-        [schema.TRACE_SPAN_NAME]: `${context.method} ${context.url.pathname}${context.url.search}`,
-        [schema.TRACE_ID_SOURCE]: traceContext.source,
-        'request.host': context.host,
-        'request.original_url': context.url.href,
-        'request.remote_addr': context.remote,
-        'request.method': context.method,
-        'request.scheme': context.url.protocol,
-        'request.path': context.url.pathname,
-        'request.query': context.url.search
-      },
-      traceContext.traceId,
-      traceContext.parentSpanId,
-      traceContext.dataset)
-
-      if (isDev()) {
-        context._honeycombTrace = trace
-      }
-
-      if (traceContext.customContext) {
-        beeline.addContext(traceContext.customContext)
-      }
-
-      if (!trace) {
-        return next(context)
-      }
-
-      const boundFinisher = beeline.bindFunctionToTrace((response: ServerResponse) => {
-        beeline.addContext({
-          'response.status_code': String(response.statusCode)
+  if (isOtel(process.env)) {
+    return function honeycombOtelTrace (next: Handler) {
+      return async (context: Context) => {
+        const tracer = getOtelTracer()
+        let carrier = {}
+
+        // Start a parent span
+        const parentSpan = tracer.startSpan(`${context.method} ${context.url.pathname}${context.url.search}`, {
+          attributes: {
+            [SemanticAttributes.HTTP_HOST]: context.host,
+            [SemanticAttributes.HTTP_URL]: context.url.href,
+            [SemanticAttributes.NET_PEER_IP]: context.remote,
+            [SemanticAttributes.HTTP_METHOD]: context.method,
+            [SemanticAttributes.HTTP_SCHEME]: context.url.protocol,
+            [SemanticAttributes.HTTP_ROUTE]: context.url.pathname,
+            [OTEL_REQ_QUERY]: context.url.search
+          }
         })
 
-        beeline.addContext({
-          'request.route': context.handler.route,
-          'request.method': context.handler.method,
-          'request.version': context.handler.version
+        // this propagator takes care of extracting trace parent
+        // and state from request headers (and so on)
+        const propagator = new W3CTraceContextPropagator()
+
+        propagator.inject(
+          otelTrace.setSpanContext(
+            ROOT_CONTEXT,
+            parentSpan.spanContext()
+          ),
+          carrier,
+          defaultTextMapSetter
+        )
+
+        // create a parent active context
+        const parentContext = propagator.extract(
+          ROOT_CONTEXT,
+          carrier,
```

Comment:
carrier needs to be the incoming request headers object in order for this to work properly :)

Reply:
also, I just realized--this needs to get moved up to like line 49, because we need to get the parentContext and then pass that into the `tracer.startSpan(` call :)